src/session.c

```diff
@@ -7,33 +7,168 @@
 /* Session stuff */
 struct p11prov_session {
     P11PROV_CTX *provctx;
+    P11PROV_SESSION_POOL *pool;
 
     CK_SLOT_ID slotid;
     CK_SESSION_HANDLE session;
 
     int refcnt;
+    int free;
 };
 
-P11PROV_SESSION *p11prov_session_new(P11PROV_CTX *ctx, CK_SLOT_ID slotid)
+struct p11prov_session_pool {
+    P11PROV_CTX *provctx;
+
+    CK_ULONG max_sessions;
+    CK_ULONG limit_sessions;
+    CK_ULONG cur_sessions;
+
+    P11PROV_SESSION **sessions;
+    int num_sessions;
```

Comment:
`num_sessions` is slightly confusing with `cur_sessions`. The other `*_sessions` seem to reflect PKCS#11 session counts. But this seems to reflect the number of `P11PROV_SESSION` structs allocated. Maybe call it `num_p11sessions` or somehow differentiate it?

Reply:
Sounds good